algs/range-doppler-terrain-correction/include/get_position.cuh

```diff
@@ -73,8 +76,124 @@ inline __device__ __host__ double ComputeSlantRangeWithLut(
     return std::sqrt(xDiff * xDiff + yDiff * yDiff + zDiff * zDiff);
 }
 
+/**
+ * Compute ground range for given slant range. Ported from SNAP's s1tbx SARGeocoding.java.
+ *
+ * @param sourceImageWidth    The source image width.
+ * @param groundRangeSpacing  The ground range spacing.
+ * @param slantRange          The slant range in meters.
+ * @param srgrCoeff           The SRGR coefficients for converting ground range to slant range.
+ *                            Here it is assumed that the polynomial is given by
+ *                            c0 + c1*x + c2*x^2 + ... + cn*x^n, where {c0, c1, ..., cn} are the SRGR coefficients.
+ * @param ground_range_origin The ground range origin.
+ * @return The ground range in meters.
+ */
+inline __device__ __host__ double ComputeGroundRange(int source_image_width, double ground_range_spacing,
+                                                     double slant_range, double* srgr_coefficients,
+                                                     int srgr_coefficients_length, double ground_range_origin) {
+    // binary search is used in finding the ground range for given slant range
+    double lower_bound = ground_range_origin;
+    const double lower_bound_slant_range =
+        math::polynomials::CalculateValue(lower_bound, srgr_coefficients, srgr_coefficients_length);
+    if (slant_range < lower_bound_slant_range) {
+        return -1.0;
+    }
+
+    double upper_bound = ground_range_origin + source_image_width * ground_range_spacing;
+    const double upper_bound_slant_range =
+        math::polynomials::CalculateValue(upper_bound, srgr_coefficients, srgr_coefficients_length);
+    if (slant_range > upper_bound_slant_range) {
+        return -1.0;
+    }
+
+    // start binary search
+    double mid_slant_range;
+    while (upper_bound - lower_bound > 0.0) {
+        const double mid = (lower_bound + upper_bound) / 2.0;
+        mid_slant_range = math::polynomials::CalculateValue(mid, srgr_coefficients, srgr_coefficients_length);
+        if (mid_slant_range < slant_range) {
+            lower_bound = mid;
+        } else if (mid_slant_range > slant_range) {
+            upper_bound = mid;
+        } else {
+            const double a = mid_slant_range - slant_range;
+            if ((a > 0 && a < 0.1) || (a <= 0.0 && 0.0 - a < 0.1)) {
+                return mid;
+            }
+        }
+    }
+
+    return -1.0;
+}
+
+/**
+ * Compute range index in source image for earth point with given zero Doppler time and slant range. Ported from SNAP's
+ * s1tbx SARGeocoding.java
+ *
+ * @param zeroDopplerTime The zero Doppler time in MJD.
+ * @param slantRange      The slant range in meters.
+ * @return The range index.
+ */
+inline __device__ __host__ double ComputeRangeIndexGrdImpl(cuda::KernelArray<SrgrCoefficientsDevice>& srgr_coefficients,
```

Comment:
Almost certainly doesn't matter in this context, but by default small structs should be passed by value:

pass by reference adds indirection -> one extra memory load

https://godbolt.org/z/jxozjvTxs